### ISSUE TITLE
use localforage in a serviceworker

### ISSUE THREAD

**Issue by remkoboschker** [mozilla/localForage#402]
Hi,

I would like to use localforage in a serviceworker. Version 1.2.4 in Chrome 43 throws a reference error on line 963 "window undefined", because there is no window in the serviceworker. 

Window was removed in #144 but reintroduced in #237. 

Should localforage or should it not be usable in a webworker / serviceworker environment?

I'll see if I can get it to work.

**Comment by tofumatt** [comment]
It is a design goal of localforage to work in web workers. You're right though ; it currently doesn't work.




If you can send a pull request that fixes it that would be great, but if not know it's something I'll eventually work on :-)



- tofumatt

**Comment by remkoboschker** [comment]
replacing .call(window) by .call(this) loads the lib just fine, but this probably introduces other issues. I'll fork and run some tests later today. Thanks.

**Comment by tofumatt** [comment]
Awesome, thanks so much.




If you run the tests using `npm test` you can run them in phantomjs, and using `grunt serve` should allow you to run them from browsers using a local node server URL.



- tofumatt

**Comment by thgreasi** [comment]
.call(typeof window !== 'undefined ? window : this)
Looks like a safe bet :-)

**Comment by remkoboschker** [comment]
:smile: sure, will use it.

**Comment by thgreasi** [comment]
[This](https://github.com/babel/babel/pull/202) might be useful.

**Comment by thgreasi** [comment]
Just took a shot on this, the web workers part, at least... Hope it will
work with service workers similarly.
Since we now deal with es6 modules, we are in "use strict" mode by default
(and babel honors this semantic). As a result `this` is undefined inside
IIFEs...

**Comment by remkoboschker** [comment]
In the service worker your fix raises
````
Uncaught (in promise) TypeError: self._initStorage is not a function at http://localhost:8080/js/lib/localforage.js:265:48
````
Also a unit test fails
````
  1) Config API Driver API "before each" hook:
     ReferenceError: Can't find variable: localforage
      at http://localhost:9999/test/test.drivers.js:11
      at http://localhost:9999/bower_components/mocha/mocha.js:4315
      at next (http://localhost:9999/bower_components/mocha/mocha.js:4626)
      at http://localhost:9999/bower_components/mocha/mocha.js:4630
      at timeslice (http://localhost:9999/bower_components/mocha/mocha.js:5761)
````
I'll work on a fix

**Comment by thgreasi** [comment]
We might just have to use a different var name for the scope. Maybe service
workers do not like redefining `self`.

**Comment by thgreasi** [comment]
Can you share your service worker test code?

**Comment by thgreasi** [comment]
Yes, I had that in mind, but wanted to avoid possible problems with the
different module bundlers/loaders.
`self` is one of the most common variable names (along with `that`) used
for sharing context with closures (it's used by localforage as well, as you
noticed). As a result, there are cases that people overwrite it without
even knowing its existence.
Anyway if have anything against my approach, I would be happy to change
that with a new commit.

**Comment by thgreasi** [comment]
Let me rephrase that, since I think I might missed my point/wasn't clear:
* For most devs/libs, `window` is the global object in JS. So for everyday webdev `window` is respencted and not redeclared. Some of them ignore the existence of `self` and even more redeclare it to pass contexts around.
* For *worker devs/libs, `self` is the global object by default. We can bet that they use it carefully, since otherwise bugs should have appear.

Still, I have no problem using just `self` if you think that all the above is just an overkill.

**Comment by remkoboschker** [comment]
@thgreasi I do not mind the extra precaution. Can you confirm that the test are passing in your localforage#webworkers-fix? I tried to copy the webworkers test to a serviceworker scenario but the tests are not passing. When I include only the test.webworker.js in the test.component.html not a single test is run.

**Comment by tofumatt** [comment]
I believe the webworker tests are commented out right now, or more accurately are marked as `skip`. You might need to change them to fix this.




Let me know if it seems like that's the case but you need more help.



—

**Comment by thgreasi** [comment]
@tofumatt, @remkoboschker 
I removed the `skip` in PR #406 and seems to work with changes I did to `call(...)` and the babel system import plugin.
I can verify that [#406 fixes web-workers and passes the test cases both on Travis](https://travis-ci.org/mozilla/localForage/builds/71530331) and locally/in-browser.

@remkoboschkerI also started looking at a service worker test case (today morning), but it is still in embryonic state I think.

**Comment by remkoboschker** [comment]
@thgreasi I have a basic unit test
serviceworker-client.js:
````
importScripts("/dist/localforage.js");

self.oninstall = function(event) {
   event.waitUntil(
       localforage
       .setItem('serviceworker', 'serviceworker present')
       .then(function(value) {
           console.log(value);
       })
   );
};
````
test.serviceworker.js:
````
describe('serviceworkers', function() {
    it('should set a value', function(done) {
       navigator
       .serviceWorker
       .register('/test/serviceworker-client.js')
       .then(function() {
           return localforage.getItem('serviceworker');
       })
       .then(function(result) {
           expect(result).to.equal('serviceworker present');
           done();
       })
       .catch(function(error) {
           done(error);
       });
   });
});
````

**Comment by thgreasi** [comment]
That looks very similar to my code! On top of that, I'm currently trying to
post the desired driver name to the service worker.

**Comment by thgreasi** [comment]
Can I find this code in any public commit that I can cherry pick and push
on top of my PR? (And also possibly cherry pick it retaining your commit
info.)

**Comment by remkoboschker** [comment]
I have pushed the branch to https://github.com/curit/localForage. In test/test.serviceworker.js en in test/serviceworker-client.js an earlier attempt to send messages is commented out. I found this example to be quite useful https://jakearchibald.github.io/isserviceworkerready/demos/postMessage/

Good luck!

**Comment by thgreasi** [comment]
Is there any chance you can rebase over the latest commit of my PR (and
even better squash some commits)?  I'm getting a huge diff on the dist
files even though I was just expecting the extra test case.
On the other hand I don't want to spam you, the service worker test case
could also be a separate PR :)

**Comment by remkoboschker** [comment]
@thgreasi sorry about the diff I was trying to get the unit test to run at all. I'm working on it. By the way how do I get at the pull request in git or should I fork your repo?

**Comment by remkoboschker** [comment]
Turns out the component-build was not being executed. So now I can run the tests and have submitted a pull request to your repo. I will now try to get the tests to pass. 

I found this useful to fetch the pull request https://help.github.com/articles/checking-out-pull-requests-locally/

**Comment by remkoboschker** [comment]
https://github.com/thgreasi/localForage/pull/1 now has both tests passing when run in Chrome 44

**Comment by remkoboschker** [comment]
and in firefox nightly with the dom.serviceworker enabled

**Comment by remkoboschker** [comment]
It does not look like phantomjs supports serviceworkers. I have asked some questions https://github.com/jakearchibald/isserviceworkerready/issues/43 and https://github.com/ariya/phantomjs/issues/13437

**Comment by thgreasi** [comment]
Just 1 thing to note: awesome :+1:

**Comment by remkoboschker** [comment]
@thgreasi Thank you. I'm away on holiday the coming weeks. So good luck with the pull request!

**Comment by thgreasi** [comment]
@remkoboschker best wishes regarding your holidays. I will try to finish this up by the time you are back.
Thanks for your collaboration.

**Comment by thgreasi** [comment]
@remkoboschker Awesome work!!! I'm speechless :open_mouth: :open_mouth: :open_mouth: :smiley: 
I will try to merge it today or at most tomorrow.

**Comment by nivle** [comment]
Does localforage still work inside a service worker??
When I try to get data from indexeddb I get null as a result even though the key value pair exists and can be gained in regular js file.
